aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -2,27 +2,46 @@
+std::tuple<Tensor&, Tensor&, Tensor&> batch_norm_cuda_out(Tensor& output, Tensor& save_mean, Tensor& save_invstd, const Tensor& self, const Tensor& weight, const Tensor& bias,
                                                    const Tensor& running_mean, const Tensor& running_var, bool train, double momentum, double epsilon) {
-  return AT_DISPATCH_FLOATING_TYPES_AND_HALF(self.scalar_type(), "batch_norm_cuda", [&] {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(self.scalar_type(), "batch_norm_cuda", [&] {
       auto mean_st = running_mean.dtype();
       auto var_st = running_var.dtype();
       TORCH_CHECK(mean_st == var_st, "running_mean and running_var need to have the same data types");
       bool is_half_float = std::is_same<scalar_t, at::Half>::value && mean_st == at::kFloat;
       if (cuda::detail::canUse32BitIndexMath(self)) {
         if (is_half_float) {
-          return batch_norm_cuda_template<at::Half, float, int32_t>(self, weight, bias, running_mean, running_var, train, momentum, epsilon);
+          batch_norm_cuda_template<at::Half, float, int32_t>(output, save_mean, save_invstd, self, weight, bias, running_mean, running_var, train, momentum, epsilon);
         } else {
-          return batch_norm_cuda_template<scalar_t, scalar_t, int32_t>(self, weight, bias, running_mean, running_var, train, momentum, epsilon);
+          batch_norm_cuda_template<scalar_t, scalar_t, int32_t>(output, save_mean, save_invstd, self, weight, bias, running_mean, running_var, train, momentum, epsilon);
         }
       } else {
         if (is_half_float) {
-          return batch_norm_cuda_template<at::Half, float, int64_t>(self, weight, bias, running_mean, running_var, train, momentum, epsilon);
+          batch_norm_cuda_template<at::Half, float, int64_t>(output, save_mean, save_invstd, self, weight, bias, running_mean, running_var, train, momentum, epsilon);
         } else {
-          return batch_norm_cuda_template<scalar_t, scalar_t, int64_t>(self, weight, bias, running_mean, running_var, train, momentum, epsilon);
+          batch_norm_cuda_template<scalar_t, scalar_t, int64_t>(output, save_mean, save_invstd, self, weight, bias, running_mean, running_var, train, momentum, epsilon);
         }
       }
     });
+  return std::tuple<Tensor&, Tensor&, Tensor&>(output, save_mean, save_invstd);
+}
+
+std::tuple<Tensor, Tensor, Tensor> batch_norm_cuda(const Tensor& self, const Tensor& weight, const Tensor& bias,
+                                                   const Tensor& running_mean, const Tensor& running_var, bool train, double momentum, double epsilon) {
+  auto output = at::empty_like(self, at::MemoryFormat::Contiguous);
+  int64_t n_input = self.size(1);
+  auto input_options = self.options();
+  Tensor save_mean, save_invstd;
+  if (train) {
+    save_mean = at::empty({n_input}, input_options);
+    save_invstd = at::empty({n_input}, input_options);
```

Comment:
The problem is here, you're allocating the accumulators at the wrong type when you're in half precision.